Add custom Elytra support

l-Luna opened the pull request:
This adds support for items to act as custom Elytras via implementing IElytra on the item. 

I know that https://github.com/MinecraftForge/MinecraftForge/pull/4476 exists, but this makes creating Elytras with custom properties significantly easier.

One use-case is a RF-powered Elytra with can speed up at the cost of more energy.

l-Luna committed:
Added stuff

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4848) <br/>All committers have signed the CLA.

ChloeDawn commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -1,6 +1,14 @@
+ import net.minecraft.world.WorldServer;
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
++
```
Imports should not be reorganized

ChloeDawn commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -31,15 +39,24 @@
+                 soundevent = ((ItemArmor)item).func_82812_d().func_185017_b();
+             }
+-            else if (item == Items.field_185160_cR)
++            else if (item == Items.field_185160_cR || item instanceof net.minecraftforge.items.IElytra)
```
Why does ItemElytra not implement this interface?

l-Luna replied:
I wanted to keep patches as small as possible, so I just added checks for IElytra alongside regular Elytra checks.

l-Luna replied:
This has now been changed

ChloeDawn commented on src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:
```diff
@@ -0,0 +1,86 @@
+	};
+	
+	
+	@Mod.EventBusSubscriber(modid = MODID)
```
There is absolutely no reason for having 2 separate event bus subscriber classes to register a test item + model. These methods can go in the main class.

l-Luna replied:
I was basing the test mod after the enchantment one

covers1624 replied:
The second one is needed for client side events, it would crash on the server otherwise as the EventBus doesnt support SideOnly

ChloeDawn commented on jsons/1.12.2.json:
```diff
@@ -719,6 +719,6 @@
     "minimumLauncherVersion": 18,
     "releaseTime": "2017-09-18T08:39:46+00:00",
-    "time": "2018-02-15T16:26:45+00:00",
+    "time": "2018-04-02T07:39:26+00:00",
```
This entire file change is unrelated to your pull request and should be excluded

l-Luna replied:
That was automatically done by gradlew genPatches

ChloeDawn replied:
That is not a valid reason for it to be included in your pull request..

l-Luna replied:
But the build time, size, and checksum _were_ changed as a result of this PR, so it should change, right?

LexManos replied:
No they went they were changed because mojang changed their assets for April fools. Don't include these jains in your prs

Leppan committed:
Update 1.12.2.json

Leppan committed:
Update 1.12.2.json

Leppan committed:
Update 1.12.2.json

Leppan committed:
Update 1.12.2.json

l-Luna committed:
Updated stuff

l-Luna committed:
Merge branch '1.12.x' of https://github.com/Leppan/MinecraftForge into 1.12.x

Leppan committed:
Update EntityLivingBase.java.patch

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -1,6 +1,13 @@
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
```
Don't touch imports in patch files.

liach replied:
Well, this patch has nothing. Weird.

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
+                 {
+                     if (this.func_184613_cA())
+                     {
++                    	ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
```
Weird spacing.

liach replied:
Please change tabs to spaces.

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
+-                            this.field_70143_R = 1.0F;
++                        	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesSlowOnUpwardsFlight(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
++                        		this.field_70143_R = 1.0F;
++                        	}
```
Tabs, space after `if`, brace on newline.

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
++                        	field_70181_x *= ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p);
++                        	field_70179_y *= ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p);
++                        }
++                        
```
Tabs, random newline, space after `if`, brace on newline.

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
++                                if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageUponHittingWall(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR) {
++                                	this.func_70097_a(DamageSource.field_188406_j, f5);
++                                	if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra) ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).onHitWall(itemstack, field_70170_p, new BlockPos(field_70165_t, field_70163_u, field_70161_v));
++                                }
```
Tabs. Space after `if`and if body on new line.

covers1624 commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -293,7 +348,24 @@
+                 {
++                	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageWhileFlying(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
++                		
++                	}
```
This if block does nothing. Also tabs, space after `if` and brace on newline.

covers1624 commented on src/main/java/net/minecraftforge/items/IElytra.java:
```diff
@@ -0,0 +1,59 @@
+package net.minecraftforge.items;
```
License header.

covers1624 commented on src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:
```diff
@@ -0,0 +1,86 @@
+package net.minecraftforge.debug.item;
```
License Header.

covers1624 commented on src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:
```diff
@@ -0,0 +1,86 @@
+		
+		@Override
+		public boolean doesSlowOnUpwardsFlight(ItemStack stack, World world)
+        {
```
This class is full of tabs, except this line.

covers1624 commented on src/main/java/net/minecraftforge/items/IElytra.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Implement this on your items to make them act as an Elytra
+ */
+public interface IElytra{
```
Brace on newline.

covers1624 commented on jsons/1.12.2.json:
```diff
@@ -1,10 +1,10 @@
+        "sha1": "e75e9535754c6f2158b0b18b35660f45c4495d78",
+        "size": 169257,
+        "url": "https://launchermeta.mojang.com/mc/assets/1.12/e75e9535754c6f2158b0b18b35660f45c4495d78/1.12.json",
+        "totalSize": 127722338
```
Revert your local changes here, if you didn't do it, don't commit it.

l-Luna replied:
These have since been changed by forge in source. The 'changes' thing only shows what it looked like as I forked it.

covers1624 reviewed (commented):
Formatting!
Whilst i see the intentions behind this, it feels extremely over complicated, Vanilla's elytra should implement the interface, and j8 defaults could be used in the interface for those optional things.

liach commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
+                         if (this.field_70181_x > -0.5D)
+                         {
+-                            this.field_70143_R = 1.0F;
++                        	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesSlowOnUpwardsFlight(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
```
Can you just let the ielytra handle field 70143 R based on field 70181 x?

liach commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -293,7 +348,24 @@
+ 
+                 if (!this.field_70170_p.field_72995_K && (this.field_184629_bo + 1) % 20 == 0)
+                 {
++                	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageWhileFlying(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
```
We shall let custom elytras determine their damage rate instead of just yes/no on damaging.

liach commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
+-                                this.func_70097_a(DamageSource.field_188406_j, f5);
++                                if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageUponHittingWall(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR) {
++                                	this.func_70097_a(DamageSource.field_188406_j, f5);
++                                	if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra) ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).onHitWall(itemstack, field_70170_p, new BlockPos(field_70165_t, field_70163_u, field_70161_v));
```
Can we just call hit wall without that boolean check? This ways modded elytras can do more things on hitting a wall.

liach commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,54 @@
+                         this.field_70159_w *= 0.9900000095367432D;
+                         this.field_70181_x *= 0.9800000190734863D;
+                         this.field_70179_y *= 0.9900000095367432D;
++                        if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p) != 1){
```
You should replace the 3 modifiders above instead of calling again. Also, I'd make a `void getAccelerationMultiplier(ItemStack, EntityLivingBase, int[] velocity)`, where the implementors just modify the int array to achieve acceleration multiplier.

pau101 replied:
I think a `Vec3d` return would be more appropriate.

liach replied:
Yes, it should be double. However I do not like vec3d for the excessive creation of unnecessary objects. (Vec3d is immutable)

pau101 replied:
An implementer doesn't need to produce a new instance each call but if it were an array you would have that instantiation (but keeping in mind JIT is also at play). However more importantly, this is the most ridiculous premature "optimization". Java is an object oriented language, use objects. This isn't the place to sacrifice API quality.

liach replied:
Then i suggest add 3 original numbers as arguments in case this elytra accelerates if one of x y z is too low. does that sound plausible?

liach reviewed (changes requested):
Why are all these parameters `ItemStack` and `World`? As I see, you can change all `World` parameters to `EntityLivingBase` ones; you can access worlds from entity instances.

l-Luna committed:
Changed stuff

l-Luna committed:
Merge branch '1.12.x' of https://github.com/Leppan/MinecraftForge into 1.12.x

l-Luna commented:
I've reformatted lots of things, changed ItemElytra to implement IElytra, changed getAccelerationModifier to changeAcceleration, onHitWall now gets called regardless, Elytras now have control over damaging rates, all IElytra methods now accept Itemstack, EntityLivingBase.

l-Luna commented:
jsons/1.12.2 has since been changed in the forge source, but the compare thing doesn't show it.

liach commented on patches/minecraft/net/minecraft/client/entity/EntityPlayerSP.java.patch:
```diff
@@ -111,3 +111,12 @@
+             ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
+ 
+-            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack))
++            if ((itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack)) || itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra)
```
IElytras should have ways to determine if the item stack can be used for flying (Like vanilla elytras can do that)

liach commented on patches/minecraft/net/minecraft/item/ItemElytra.java.patch:
```diff
@@ -0,0 +1,24 @@
+         }
+     }
++    // FORGE START
++    public boolean doesSlowOnUpwardsFlight(ItemStack stack, EntityLivingBase entity){return true;}
```
Add `@Override`s

liach commented on src/main/java/net/minecraftforge/items/IElytra.java:
```diff
@@ -0,0 +1,82 @@
+     * @param entity
+     *            The user
+     */
+    public void changeAcceleration(ItemStack stack, EntityLivingBase entity, double motionX, double motionY, double motionZ);
```
The return type should be Vec3d

liach commented on src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:
```diff
@@ -0,0 +1,91 @@
+            // Goes a bit faster than normal. This escalates QUICKLY; it can be changed as you use it, but I'd recommend sticking to 1 for most things.
+            motionX *= 1.1;
+            motionY *= 1.1;
+            motionZ *= 1.1;
```
Uhh this does not work.
